content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,275 @@
+
+2. **Burstable pods** by their QoS definition require at least one container in the Pod with
+CPU or memory request or limit set.
+    
+    * When requests.memory and limits.memory are set, the formula is used as-is:
+    ![](./container-memory-high-limit.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+    * When requests.memory is set, limits.memory is not set, we substitute limits.memory
+    for node allocatable memory in the formula:
+    ![](./container-memory-high-no-limits.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+3. **BestEffort** by their QoS definition do not require any memory or CPU limits or requests.
+For this case, we set requests.memory = 0 and substitute limits.memory for node allocatable
+memory in the formula:
+
+    ![](./container-memory-high-best-effort.svg)  
+    <sub>i: the i<sup>th</sup> container in one pod</sub>
+
+**Summary**: Only Pods in Burstable and Best effort QoS classes will set memory.high.
+Guaranteed QoS pods do not set memory.high as their memory is guaranteed.
+
+### `memory.min` calculations for cgroup heirarchy
+
+When container memory requests are made, kubelet passes `memory.min` to the back-end 
+CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 
+container creation. The `memory.min` in container level cgroup will be set to:  
+
+$memory.min =  pod.spec.containers[i].resources.requests[memory]$  
+<sub>for every i<sup>th</sup> container in a pod</sub>
+
+Since the `memory.min` interface requires that the ancestor cgroup directories are all 
+set, the pod and node cgroup directories need to be set correctly. 
+
+`memory.min` in pod level cgroup:  
+$memory.min = \sum_{i=0}^{no. of pods}pod.spec.containers[i].resources.requests[memory]$  
+
+`memory.min` in node level cgroup:  
+$memory.min = \sum_{i}^{no. of nodes}\sum_{j}^{no. of pods}pod[i].spec.containers[j].resources.requests[memory]$  
+
+Kubelet will manage the cgroup hierarchy of the pod level and node level cgroups 
+directly using the libcontainer library (from the runc project), while container
+cgroup limits are managed by the container runtime.
+
+## How do I use it?
+
+The prerequisites for enable Memory QoS feature on your Linux node are:
+
+1. Verify the [requirements](/docs/concepts/architecture/cgroups/#requirements) 
+related to [Kubernetes support for cgroup v2](/docs/concepts/architecture/cgroups)
+are met. 
+2. Ensure CRI Runtime supports Memory QoS. In Alpha phase of this feature,
+only containerd and cri-o are supported. Related PRs:
+    * Containerd: [Feature: containerd-cri support LinuxContainerResources.Unified #5627](https://github.com/containerd/containerd/pull/5627).
+    * CRI-O: [implement kube alpha features for 1.22 #5207](https://github.com/cri-o/cri-o/pull/5207).
+
+Memory QoS is in alpha for Kubernetes v1.27. You can enable the feature by setting
+the command line arguments:
+
+```
+--feature-gates="...,MemoryQoS=true"
```

Comment:
For the kubelet, setting `--feature-gates` is **deprecated** and we therefore shouldn't recommend it. Instead, suggest editing the kubelet config file.